### PR TITLE
Quick settings tile for Android version 7.0+

### LINKED ISSUE #26: Add a toggle button in the pull-down shade/menu for android 7.0+
Add a toggle button in the pull-down shade/menu for android 7.0+ (similar to how PIA app does)

### PR DESCRIPTION
**Pull request by calimbak**
Adds a tile in the quick settings with a basic icon and text for now.
The tile acts as a toggle starting / stopping the AdVpnService.
Update of build versions and dependencies to be able to use the tile.

Fixes #26

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -1,12 +1,12 @@
...
 android {
-    compileSdkVersion 24
-    buildToolsVersion "24.0.3"
+    compileSdkVersion 25
+    buildToolsVersion "25.0.0"
     defaultConfig {
         applicationId "org.jak_linux.dns66"
         minSdkVersion 21
-        targetSdkVersion 24
+        targetSdkVersion 25
```
Comment (julian-klode) [inline app/build.gradle]: I don't see why the changes are necessary. The tile API was introduced in level 24 according to the documentation. And if they really are, they should be in a separate commit prior to the one adding the functionality.

File: app/build.gradle [hunk 2]
```diff
@@ -27,10 +27,10 @@ dependencies {
...
         exclude group: 'com.android.support', module: 'support-annotations'
     })
-    compile 'com.android.support:appcompat-v7:24.2.1'
-    compile 'com.android.support:support-v4:24.2.1'
-    compile 'com.android.support:design:24.2.1'
-    compile 'com.android.support:recyclerview-v7:24.2.1'
+    compile 'com.android.support:appcompat-v7:25.1.0'
+    compile 'com.android.support:support-v4:25.1.0'
+    compile 'com.android.support:design:25.1.0'
+    compile 'com.android.support:recyclerview-v7:25.1.0'
```
Comment (julian-klode) [inline app/build.gradle]: Same for all these changes.

File: app/src/main/java/org/jak_linux/dns66/MainActivity.java [hunk 3]
```diff
@@ -170,7 +170,7 @@ private void refresh() {
 
             if (file != null && item.state != 2) {
                 DownloadManager.Request request = new DownloadManager.Request(Uri.parse(item.location));
-                Log.d("MainActivity", String.format("refresh: Downkoading %s to %s", item.location, file.getAbsolutePath()));
+                Log.d("MainActivity", String.format("refresh: Downloading %s to %s", item.location, file.getAbsolutePath()));
```
Comment (julian-klode) [inline app/src/main/java/org/jak_linux/dns66/MainActivity.java]: Split into a typo fix commit.

File: app/src/main/java/org/jak_linux/dns66/main/StartFragment.java [hunk 4]
```diff
@@ -115,12 +119,20 @@ private void startService() {
...
+        getActivity().startService(intent);
+    }
+
     /**
      * Check if all configured hosts files exist.
      *
      * @return true if all host files exist or no host files were configured.
      */
-    private boolean areHostsFilesExistant() {
+    private boolean areHostsFilesExistent() {
```
Comment (julian-klode) [inline app/src/main/java/org/jak_linux/dns66/main/StartFragment.java]: split into typo fix commit

File: app/src/main/java/org/jak_linux/dns66/main/StartFragment.java [hunk 5]
```diff
@@ -105,8 +101,16 @@ public void onClick(DialogInterface dialog, int which) {
...
+	/**
+     * Called when the start of the service was canceled after some host files were missing
+     */
+    protected void onCancelStart() {
+         /* Do nothing */
+    }
+
     private void startService() {
         Log.i(TAG, "Attempting to connect");
+
```
Comment (julian-klode) [inline app/src/main/java/org/jak_linux/dns66/main/StartFragment.java]: drop that change, it's unrelated.

File: app/src/main/java/org/jak_linux/dns66/main/StartFragment.java [hunk 6]
```diff
@@ -81,16 +77,16 @@ public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
...
+        if (!areHostsFilesExistent()) {
             new AlertDialog.Builder(getActivity())
                     .setIcon(R.drawable.ic_warning)
                     .setTitle(R.string.missing_hosts_files_title)
                     .setMessage(R.string.missing_hosts_files_message)
                     .setNegativeButton(R.string.button_no, new DialogInterface.OnClickListener() {
                         @Override
                         public void onClick(DialogInterface dialog, int which) {
-                            /* Do nothing */
+                            onCancelStart();
```
Comment (julian-klode) [inline app/src/main/java/org/jak_linux/dns66/main/StartFragment.java]: Do the method extraction in a separate commit.

File: app/src/main/java/org/jak_linux/dns66/main/StartFragment.java [hunk 7]
```diff
@@ -57,11 +57,7 @@ public View onCreateView(LayoutInflater inflater, ViewGroup container,
             @Override
             public boolean onLongClick(View v) {
                 if (AdVpnService.vpnStatus != AdVpnService.VPN_STATUS_STOPPED) {
-                    Log.i(TAG, "Attempting to disconnect");
-
-                    Intent intent = new Intent(getActivity(), AdVpnService.class);
-                    intent.putExtra("COMMAND", org.jak_linux.dns66.vpn.Command.STOP.ordinal());
-                    getActivity().startService(intent);
+                    stopService();
```
Comment (julian-klode) [inline app/src/main/java/org/jak_linux/dns66/main/StartFragment.java]: Do the method extraction in a separate commit.

**Comment by julian-klode** [comment]
Sorry it took so long to review, I thought I had send the review last week, but apparently did not.

**Comment by calimbak** [comment]
I decided to use the new Activity and Fragment to reuse the existing implementation of `StartFragment` and avoid having any duplication of code. Since we also need the result of `startActivityForResult` when preparing the vpn service (which cannot be received in a Service), it seems like the best way to do it. 

Oh yes, I will separate everything in different commits.

I agree with you for the icon, was just using the power icon as a placeholder so that the icon can just be replaced when we have the correct one.

File: app/build.gradle (hunk 1 above)
Comment (calimbak) [inline app/build.gradle]: Don't know what I was thinking when I changed that, but you are right we don't need to change the Android version.

**Comment by ashreyas** [comment]
Is this not included in Release 0.6.5?
I can't find any tile related to DNS66. I'm using Android 10.

**Comment by julian-klode** [comment]
Closing due to inactivity.

**Comment by schirinowski** [comment]
@julian-klode, does this mean the QuickSettings tile has been entirely discarded?
It really would be handy to have it. It's already great to have the notification, but it needs one tap and often one slide more to toggle DNS66.
